### DISCUSSION TITLE
Sound Markers

### DISCUSSION THREAD

**Discussion by ankushdineshrana** [ggerganov/ggwave#13] [Q&A] [answered]
Hi Georgi,

First of all kudos to you for bringing such an amazing work. I myself tried various publicly available open source offerings on Data-Over-Sound. But its only **_ggwave_** that works seamlessly.

Now comes the point of discussion:

Georgi, I read that you are using sound markers to record/analyze the sound captured between them. Can you put some more light on it as in how it actually works in your code and can we customize it as per our needs?

Thanks a lot.

Regards,
Ankush

**Comment by ggerganov** [comment, marked as answer]
Hi Ankush,

First - thank you very much for the feedback! Such type of information is very useful to me since I am only capable of testing `ggwave` on a limited amount of devices and environments. Knowing if it works or not for other people will help to improve the robustness of the protocol.

The sound markers currently used in `ggwave` are the main thing I would like to potentially change in the future. The main reason is that I am not happy with how they currently sound for audible protocols. Still, I think they are quite robust and for now they seem to work pretty OK.

The main purpose of the sound markers is to provide a cheap (in terms of CPU) way to determine if there is an incoming audio message. As audio samples are being continuously captured, `ggwave` looks for **begin** and **end** markers and records everything in-between so it can be analyzed after the recording is completed. The **begin** and **end** sound markers are slightly different.

The **begin** marker uses the first 32 frequency bins: <tt>F0 + 00\*df</tt> to <tt>F0 + 31\*df</tt>. We produce a waveform which consists of 16 frequencies:
- <tt>F0 + 00*df</tt>
- <tt>F0 + 03*df</tt>
- <tt>F0 + 04*df</tt>
- <tt>F0 + 07*df</tt>
- <tt>F0 + 08*df</tt>
- <tt>F0 + 11*df</tt>
- <tt>F0 + 12*df</tt>
- <tt>F0 + 15*df</tt>
- <tt>F0 + 16*df</tt>
- <tt>F0 + 19*df</tt>
- <tt>F0 + 20*df</tt>
- <tt>F0 + 23*df</tt>
- <tt>F0 + 24*df</tt>
- <tt>F0 + 27*df</tt>
- <tt>F0 + 28*df</tt>
- <tt>F0 + 31*df</tt>

https://github.com/ggerganov/ggwave/blob/2811934e45c0f1cf866f0d71e038a35e91cdfa42/src/ggwave.cpp#L435-L444

The **end** marker consists again of 16 frequencies - the ones that are missing in the **begin** marker:

https://github.com/ggerganov/ggwave/blob/2811934e45c0f1cf866f0d71e038a35e91cdfa42/src/ggwave.cpp#L492-L499

The purpose of encoding the markers in this way is that one can easily detect the markers. For each frame of input samples, compute the FFT and analyze the 32 bins. To detect a **begin** marker:

https://github.com/ggerganov/ggwave/blob/2811934e45c0f1cf866f0d71e038a35e91cdfa42/src/ggwave.cpp#L723-L738

And to detect an **end** marker:

https://github.com/ggerganov/ggwave/blob/2811934e45c0f1cf866f0d71e038a35e91cdfa42/src/ggwave.cpp#L759-L777

This detection is robust because we are comparing the relative strength of two neighboring frequencies to determine if we have a **begin** or **end** marker. So no background noise estimation is necessary.

We emit the markers for certain amount of time - currently 16 frames. The shorter amount of time - the lower probability to detect the marker. So it is a balance between robustness and speed.

If we use less than 16 frequencies then again - robustness is reduced because a random noise fluctuation can accidentally produce a **begin** marker. I found that 16 frequencies behaves good enough to have a small number of false positives.

It is possible to make the marker parameters part of the protocol parameters so that the user can configure to some extend the markers. But as I mentioned, I want to find some other way to create encode the markers so that the audio is less annoying, so that is way I am still hesitating to do that.

**Reply by ankushdineshrana** [reply]
True that, sound markers performed robustly during my trials as well. Having a configurable markers could really be a nice feature. Thanks for your detailed answer @ggerganov . This will definitely help me to develop a better understanding now.
